node/node.go

```diff
@@ -311,6 +328,37 @@ func (n *Node) addConn(c *Conn) {
 	n.conns = append(n.conns, c)
 }
 
+func (n *Node) updateServiceDiscovery(conn *factory.Connection) {
+	feeds := n.Feeds()
+	services := make([]*factory.Service, len(feeds))
+	for i, feed := range feeds {
+		services[i] = &factory.Service{Key: feed}
+	}
+	conn.FindServiceNodesByKeys(feeds)
+	if n.conf.PublicServer {
+		conn.UpdateServices(&factory.NodeServices{ServiceAddress: n.conf.Listen, Services: services})
+	}
+}
+
+func (n *Node) findServiceNodesCallback(resp *factory.QueryResp) {
+	if len(resp.Result) < 1 {
+		return
+	}
+	for k, v := range resp.Result {
+		key, err := cipher.PubKeyFromHex(k)
+		if err != nil {
+			continue
+		}
+		for _, addr := range v {
+			err = n.Connect(addr)
+			if err != nil {
+				continue
+			}
+			n.Connection(addr).Subscribe(key)
```

Comment:
I dont think so. 
I have read your code. A `Connection` call will not return `nil` after a successful A `Connect` call.
Because after `Connect` is called without err, a *Conn will put into map p.conns 
```go
func (p *Pool) Connection(address string) *Conn {
	p.cmx.Lock()
	defer p.cmx.Unlock()

	return p.conns[address]
}
```
if `n.Connect(addr)` return `nil` not an error, 
then I think the `Connect` call is fine, and n.Connection(addr) would not return `nil`.
otherwise, what do you think about it when you design it?
```go
		for _, addr := range v {
			err = n.Connect(addr)
			if err != nil {
				continue
			}
			n.Connection(addr).Subscribe(key)
		}
```
and node.Config.OnCreateConenction callback I think is designed and used for app like bbs, not in this situation, and if so, it will create problems.

Reply:
```go
func (p *Pool) Connection(address string) *Conn {
	p.cmx.Lock()
	defer p.cmx.Unlock()

	return p.conns[address]
}
```
Now we are using `node.Conn` instead of `gnet.Conn` and the `node.Conn` created after successful handshake. E.g. the `(*Pool) Connect` is synchronous. Yes it's true. But `node.Connect` is not.

Thus, there are two ways:
1) I can make it synchronous. This way, the `Connect` will return `(c *Conn, err error)`

Reply:
And another way

Reply:
What is the other way?
I think it will panics in here!
```go
func (s *Node) Connection(address string) (c *Conn) {
	if gc := s.pool.Connection(address); gc != nil {
		c, _ = gc.Value().(*Conn)
	}
	return
}
```
I think it is weird here, and should be improved.

Reply:
I think instead of return nil, it will panic here:
c, _ = gc.Value().(*Conn)

Reply:
It will not panic.

Reply:
It returns nil, if handshake has not been performed.

Reply:
The `c, _ = gc.Value().(*Conn)` is type assertion with "ok". Take a look https://tour.golang.org/methods/15

Reply:
Connecting steps:
1. create `gnet.Conn`
2. perform handshake
3. add add `node.Conn` to the `gnet.Conn.Value`. add the `node.Conn` to list of connections.
4. Profit!
And if the handshake fails (by timeout for example), then step (3) will never be reached. And the `gnet.Conn` will be closed.

Reply:
No worries. I think it's better to make it synchronous. Any other way will be more complicated.

Reply:
lol, I thought . on a nil will panic at first look.
But I remembered after a while.